flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannelTest.java

```diff
@@ -306,51 +306,88 @@ public void testProducerFailedException() throws Exception {
 
-			inputChannel.onSenderBacklog(11);
+			// Increase the backlog to exceed the number of available floating buffers
+			inputChannel.onSenderBacklog(10);
 
-			// Need extra three floating buffers, but only two buffers available in buffer pool, register as listener as a result
-			verify(bufferPool, times(11)).requestBuffer();
+			// The channel does not get enough floating buffer and register as buffer listener
+			verify(bufferPool, times(13)).requestBuffer();
 			verify(bufferPool, times(1)).addBufferListener(inputChannel);
-			assertEquals("There should be 12 buffers available in the channel",
-				12, inputChannel.getNumberOfAvailableBuffers());
+			assertEquals("There should be 11 buffers available in the channel", 11, inputChannel.getNumberOfAvailableBuffers());
+			assertEquals("There should be 12 buffers required in the channel", 12, inputChannel.getNumberOfRequiredBuffers());
+			assertEquals("There should be 0 buffer available in local pool", 0, bufferPool.getNumberOfAvailableMemorySegments());
 
-			inputChannel.onSenderBacklog(12);
+			// Continue increasing the backlog
+			inputChannel.onSenderBacklog(11);
 
-			// Already in the status of waiting for buffers and will not request any more
-			verify(bufferPool, times(11)).requestBuffer();
+			// The channel is already in the status of waiting for buffers and will not request any more
+			verify(bufferPool, times(13)).requestBuffer();
 			verify(bufferPool, times(1)).addBufferListener(inputChannel);
+			assertEquals("There should be 11 buffers available in the channel", 11, inputChannel.getNumberOfAvailableBuffers());
+			assertEquals("There should be 13 buffers required in the channel", 13, inputChannel.getNumberOfRequiredBuffers());
+			assertEquals("There should be 0 buffer available in local pool", 0, bufferPool.getNumberOfAvailableMemorySegments());
+
+			// Recycle the floating buffer and assign it to the buffer listener
+			floatingBuffer1.recycle();
+
+			// The channel is still waiting for one more floating buffer
+			assertEquals("There should be 12 buffers available in the channel", 12, inputChannel.getNumberOfAvailableBuffers());
+			assertEquals("There should be 13 buffers required in the channel", 13, inputChannel.getNumberOfRequiredBuffers());
+			assertEquals("There should be 0 buffer available in local pool", 0, bufferPool.getNumberOfAvailableMemorySegments());
+
+			// Recycle one more floating buffer again
+			floatingBuffer2.recycle();
+
+			// The channel already gets all the required buffers
+			assertEquals("There should be 13 buffers available in the channel", 13, inputChannel.getNumberOfAvailableBuffers());
+			assertEquals("There should be 13 buffers required in the channel", 13, inputChannel.getNumberOfRequiredBuffers());
+			assertEquals("There should be 0 buffer available in local pool", 0, bufferPool.getNumberOfAvailableMemorySegments());
+
```

Comment:
Whenever `inputChannel.getNumberOfAvailableBuffers()` reaches `inputChannel.getNumberOfRequiredBuffers()`, we should not listen to / add new buffers. We should verify this behaviour and it could result from three things:
1. enough floating buffers become available
2. enough exclusive buffers become available
3. the sender backlog decreases

We can verify that we do not listen to new buffers anymore indirectly by recycling yet another floating buffer and verifying that `inputChannel.getNumberOfAvailableBuffers()` does not change.

Can you add these two additional checks?

Reply:
These new additional checks help me find an incorrect point in `RemoteInputChannel#notifyBufferAvailable`. Thanks for the suggestions!

Reply:
Great. I'm unsure, however, whether you tested all three cases now, i.e. getting from a state of `getNumberOfRequiredBuffers() > getNumberOfAvailableBuffers()` to a state of equality via all three means. Your last tests all come from `getNumberOfRequiredBuffers() <= getNumberOfAvailableBuffers()` only for which you check the behaviour. Also, `testRequestAndReturnFloatingBuffer()` is getting quite long and complex now.

Should we have those as (three?) separate tests instead (and cover these two origins one after another there, starting with `getNumberOfRequiredBuffers() > getNumberOfAvailableBuffers()` first, getting to `getNumberOfRequiredBuffers() = getNumberOfAvailableBuffers()` via one of the three causes, and getting to `getNumberOfRequiredBuffers() <= getNumberOfAvailableBuffers()` via one of them, too)